enigma/src/main/java/org/quiltmc/enigma/impl/source/vineflower/EnigmaTextTokenCollector.java

```diff
@@ -80,73 +108,273 @@ public void addTokensToIndex(SourceIndex index, UnaryOperator<Token> tokenProces
+				if (decl.getNameAsString().equals(parts[0])) {
+					type = decl;
+					break;
+				}
+			}
+
+			for (int i = 1; i < parts.length; i++) {
+				if (type != null) {
+					TypeDeclaration<?> finalType = type;
+					String name = parts[i];
+					type = type.findFirst(TypeDeclaration.class, t -> t != finalType && t.getNameAsString().equals(name)).orElse(null);
+				}
+			}
+
+			if (type == null) {
+				throw new IllegalStateException("Could not find type " + entry.getContextualName() + " in parsed source");
+			}
+
+			Map<String, TextRange> lambdaRanges = new HashMap<>();
+			int lambdaOrdinal = 0;
+			for (var member : type.getMembers()) {
+				if (member instanceof TypeDeclaration<?>) {
+					continue;
+				}
+
+				List<LambdaExpr> lambdas = member.findAll(LambdaExpr.class);
+				for (LambdaExpr lambda : lambdas) {
+					Node node = lambda;
+					String context;
+					while (true) {
+						Optional<Node> nodeOptional = node.getParentNode();
+						if (nodeOptional.isEmpty()) {
+							throw new IllegalStateException("Lambda has no valid parent");
+						}
+
+						node = nodeOptional.get();
+						if (node instanceof MethodDeclaration method) {
+							context = method.getNameAsString();
+							break;
+						} else if (node instanceof ConstructorDeclaration) {
+							context = "new";
+							break;
+						} else if (node instanceof TypeDeclaration<?>) {
+							context = "static";
+							break;
+						}
+					}
+
+					String name = "lambda$" + context + "$" + lambdaOrdinal++;
+					Range range = lambda.getRange().orElseThrow(() -> new IllegalStateException("No range for lambda"));
+					int start = this.positionToIndex(range.begin);
+					int end = this.positionToIndex(range.end);
+					TextRange textRange = new TextRange(start, end - start);
+					lambdaRanges.put(name, textRange);
+				}
+			}
+
+			StructClass clazz = DecompilerContext.getStructContext().getClass(entry.getFullName());
+			for (StructMethod method : clazz.getMethods()) {
+				TextRange range = lambdaRanges.get(method.getName());
+				if (range == null) {
+					continue;
+				}
+
+				SyntheticMethodSpan span = new SyntheticMethodSpan(range, true);
+				this.syntheticMethods.add(span);
+				this.syntheticEntryBySpan.put(span, getMethodEntry(entry.getFullName(), method.getName(), MethodDescriptor.parseDescriptor(method.getDescriptor())));
+			}
+		}
+	}
+
+	private void addClassAndChildren(TypeDeclaration<?> decl, String name) {
+		Range range = decl.getRange().orElseThrow(() -> new IllegalStateException("No range for type declaration"));
+		TextRange textRange = new TextRange(this.positionToIndex(range.begin), this.positionToIndex(range.end));
+		this.classRanges.put(getClassEntry(name), textRange);
+		decl.getMembers().forEach(member -> {
+			if (member instanceof TypeDeclaration<?> child) {
+				this.addClassAndChildren(child, name + "$" + child.getNameAsString());
+			}
+		});
+	}
+
+	private int positionToIndex(Position position) {
```

Comment:
I think it'd be better to greedily build a `List<Integer> indexesByLine` in `parseSource`.